File: c/aoc_2025_02.c
#include <stdint.h>
#include <stdlib.h>
#include <stdbool.h>
#include <inttypes.h>

typedef long long ll;

static inline int digits_count(int64_t x) {
    if (x <= 0) return 1;
    int d = 0;
    while (x > 0) { x /= 10; d++; }
    return d;
}

static inline int64_t pow10i(int n) {
    int64_t r = 1;
    for (int i = 0; i < n; ++i) r *= 10;
    return r;
}

ll p1(ll* input, int count) {
    ll result128 = 0;
    for (int i = 0; i < count; ++i) {
        int64_t start = (int64_t)input[2*i];
        int64_t end = (int64_t)input[2*i+1];
        if (end < 10) continue;
        int maxLen = digits_count(end);
        int maxHalf = maxLen / 2;
        for (int half = 1; half <= maxHalf; ++half) {
            int64_t pow = pow10i(half);
            int64_t scale = pow + 1;
            int64_t minFirst = pow10i(half - 1);
            int64_t maxFirst = pow - 1;
            int64_t low = start / scale;
            if (start % scale != 0) low += 1;
            if (low < minFirst) low = minFirst;
            int64_t high = end / scale;
            if (high > maxFirst) high = maxFirst;
            if (low > high) continue;

            ll cnt = (high - low) + 1;
            ll sumFirst = (low + high) * cnt / 2;
            ll add = scale * sumFirst;
            result128 += add;
        }
    }
    return result128;
}

ll p2(ll* input, int count) {
    ll result128 = 0;
    for (int i = 0; i < count; ++i) {
        int64_t start = (int64_t)input[2*i];
        int64_t end = (int64_t)input[2*i+1];
        if (end < 10) continue;
        int maxLen = digits_count(end);

        size_t seenCap = 64;
        size_t seenLen = 0;
        int64_t *seen = (int64_t*)malloc(sizeof(int64_t) * seenCap);
        if (!seen) return 0; // out-of-memory guard

        for (int L = 1; L <= maxLen / 2; ++L) {
            int64_t powL = pow10i(L);
            int kMax = maxLen / L;
            for (int k = 2; k <= kMax; ++k) {
                int totalLen = L * k;
                ll powTotal = 1;
                for (int t = 0; t < totalLen; ++t) powTotal *= 10;
                ll powL128 = powL;
                if (powL128 - 1 == 0) continue;
                ll scale128 = (powTotal - 1) / (powL128 - 1);
                if (scale128 <= 0) continue;
                if (scale128 > INT64_MAX) continue;
                int64_t scale = (int64_t)scale128;

                int64_t minFirst = pow10i(L - 1);
                int64_t maxFirst = powL - 1;
                int64_t low = start / scale;
                if (start % scale != 0) low += 1;
                if (low < minFirst) low = minFirst;
                int64_t high = end / scale;
                if (high > maxFirst) high = maxFirst;
                if (low > high) continue;

                for (int64_t first = low; first <= high; ++first) {
                    ll num128 = first * scale128;
                    if (num128 < start || num128 > end) continue;
                    int64_t num = (int64_t)num128;
                    bool found = false;
                    for (size_t j = 0; j < seenLen; ++j) {
                        if (seen[j] == num) { found = true; break; }
                    }
                    if (!found) {
                        if (seenLen >= seenCap) {
                            seenCap *= 2;
                            int64_t *tmp = (int64_t*)realloc(seen, sizeof(int64_t) * seenCap);
                            if (!tmp) { free(seen); return 0; }
                            seen = tmp;
                        }
                        seen[seenLen++] = num;
                    }
                }
            }
        }

        for (size_t j = 0; j < seenLen; ++j) result128 += seen[j];
        free(seen);
    }
    return result128;
}

File: aoc_2025_02.ts
import z from "zod";
import { fetchInput } from "./session";

const parse = (input: string) =>
	z
		.array(z.tuple([z.number(), z.number()]))
		.parseAsync(input.split(",").map((x) => x.split("-").map(Number)));
const _input = await fetchInput().then(parse);
const _example = await parse(
	`11-22,95-115,998-1012,1188511880-1188511890,222220-222224,
	1698522-1698528,446443-446449,38593856-38593862,565653-565659,
	824824821-824824827,2121212118-2121212124
	`,
);
export const p1ex = () => p1(_example);
export const p1 = (input = _input) => {
	let result = 0;
	for (const [start, end] of input) {
		if (end < 10) continue;
		const maxLen = Math.floor(Math.log10(end)) + 1;
		const maxHalf = Math.floor(maxLen / 2);
		for (let half = 1; half <= maxHalf; half++) {
			const pow = 10 ** half;
			const scale = pow + 1;
			const minFirst = 10 ** (half - 1);
			const maxFirst = pow - 1;
			const low = Math.max(minFirst, Math.ceil(start / scale));
			const high = Math.min(maxFirst, Math.floor(end / scale));
			for (let first = low; first <= high; first++) {
				result += first * scale;
			}
		}
	}
	return result;
};
export const p2ex = () => p2(_example);
export const p2 = async (input = _input) => {
	let result = 0;
	for (const [start, end] of input) {
		const seen = new Set<number>();
		if (end < 10) continue;
		const maxLen = Math.floor(Math.log10(end)) + 1;
		for (let L = 1; L <= Math.floor(maxLen / 2); L++) {
			const powL = 10 ** L;
			const kMax = Math.floor(maxLen / L);
			for (let k = 2; k <= kMax; k++) {
				const totalLen = L * k;
				const powTotal = 10 ** totalLen;
				const scale = (powTotal - 1) / (powL - 1);
				const minFirst = 10 ** (L - 1);
				const maxFirst = powL - 1;
				const low = Math.max(minFirst, Math.ceil(start / scale));
				const high = Math.min(maxFirst, Math.floor(end / scale));
				for (let first = low; first <= high; first++) {
					const num = first * scale;
					if (num >= start && num <= end) seen.add(num);
				}
			}
		}
		for (const v of seen) result += v;
	}
	return result;
};

import { cc, ptr } from "bun:ffi";
// @ts-expect-error
import source from "./c/aoc_2025_02.c" with { type: "file" };

const lib = cc({
	source,
	symbols: {
		p1: {
			args: ["ptr", "int"],
			returns: "i64",
		},
		p2: {
			args: ["ptr", "int"],
			returns: "i64",
		},
	},
});

const parseToBuffer = (input = _input) => {
	const bigints = new BigInt64Array(input.flat().map(BigInt));
	return { ptr: ptr(bigints), count: input.length };
};

export const p1_c = ({ ptr, count } = parseToBuffer()) =>
	Number(lib.symbols.p1(ptr, count));

export const p2_c = ({ ptr, count } = parseToBuffer()) =>
	Number(lib.symbols.p2(ptr, count));
